src/watchdog.js

```diff
@@ -0,0 +1,340 @@
+			.then( editor => {
+				this._editor = editor;
+
+				window.addEventListener( 'error', this._boundErrorWatcher );
+				this.listenTo( editor.model.document, 'change:data', this._throttledSave );
+
+				this._lastDocumentVersion = editor.model.document.version;
+				this._data = editor.getData();
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Destroys the current editor.
+	 *
+	 * @returns {Promise}
+	 */
+	destroy() {
+		window.removeEventListener( 'error', this._boundErrorWatcher );
+		this.stopListening( this._editor.model.document, 'change:data', this._throttledSave );
+
+		return Promise.resolve()
+			.then( () => this._destructor( this._editor ) )
+			.then( () => {
+				this._editor = null;
+			} );
+	}
+
+	/**
+	 * @private
+	 */
+	_save() {
+		const version = this._editor.model.document.version;
+
+		// Change may not produce an operation, so the document's version
+		// can be the same after that change.
+		if ( version < this._lastDocumentVersion ) {
+			this._throttledSave.cancel();
+
+			return;
+		}
+
+		try {
+			this._data = this._editor.getData();
+			this._lastDocumentVersion = version;
+		} catch ( err ) {
+			console.error(
+				err,
+				'An error happened during restoring editor data. ' +
+				'Editor will be restored from the previously saved data.'
+			);
+		}
+	}
+
+	/**
+	 * @private
+	 * @param {ErrorEvent} event
+	 */
+	_watchForEditorErrors( event ) {
+		if ( !event.error.is || !event.error.is( 'CKEditorError' ) ) {
+			return;
+		}
+
+		if ( !event.error.ctx ) {
+			console.error( 'The error is missing its context and Watchdog cannot restart the proper editor.' );
+
+			return;
+		}
+
+		if ( this._isErrorComingFromThisEditor( event.error ) ) {
+			this.crashes.push( {
+				message: event.error.message,
+				source: event.source,
+				lineno: event.lineno,
+				colno: event.colno
+			} );
+
+			this.fire( 'error' );
+
+			if ( this.crashes.length <= this._crashNumberLimit ) {
+				this.restart();
+			}
+		}
+	}
+
+	/**
+	 * @private
+	 * @param {module:utils/ckeditorerror~CKEditorError} error
+	 */
+	_isErrorComingFromThisEditor( error ) {
+		return (
+			areElementsConnected( this._editor, error.ctx ) ||
+			areElementsConnected( error.ctx, this._editor )
+		);
+	}
+}
+
+mix( Watchdog, EmitterMixin );
+
+// Returns `true` when the second parameter can be found from the first by walking through
+// the first argument.
+function areElementsConnected( from, searchedElement ) {
+	const nodes = [ from ];
+
+	// Elements are stored to prevent infinite looping.
+	const storedElements = new WeakSet();
+
+	while ( nodes.length > 0 ) {
+		// BFS should be faster.
+		const node = nodes.shift();
+
+		if ( node === searchedElement ) {
+			return true;
+		}
+
+		if ( storedElements.has( node ) || shouldNodeBeSkipped( node ) ) {
+			continue;
+		}
+
+		storedElements.add( node );
+
+		if ( Array.isArray( node ) ) {
+			nodes.push( ...node );
+		} else {
+			nodes.push( ...Object.values( node ) );
```

Comment:
Maybe we should handle also `Map`s and `Set`s here since we use a lot of them in our code?

Reply:
I'm not sure if it's needed. But actually, there should be a very easy fix for it.